clouddrift/analysis.py

```diff
@@ -1,7 +1,94 @@
 import numpy as np
-from typing import Optional, Tuple
+from collections.abc import Iterable
+from typing import Optional, Tuple, Union
 import xarray as xr
+from concurrent import futures
 from clouddrift.haversine import distance, bearing
+from clouddrift.dataformat import unpack_ragged
+
+
+def apply_ragged(
+    func: callable,
+    arrs: list[np.ndarray],
+    rowsize: list[int],
+    *args: tuple,
+    max_workers: int = None,
+    **kwargs: dict,
+) -> Union[tuple[np.ndarray], np.ndarray]:
+    """Apply a function to a ragged array.
+
+    The function ``func`` will be applied to each contiguous row of ``arrs`` as
+    indicated by row sizes ``rowsize``. The output of ``func`` will be
+    concatenated into a single ragged array.
+
+    This function uses ``concurrent.futures.ThreadPoolExecutor`` to run ``func``
+    in multiple threads. The number of threads can be controlled by the
+    ``max_workers`` argument, which is passed down to ``ThreadPoolExecutor``.
+
+    Parameters
+    ----------
+    func : callable
+        Function to apply to each row of each ragged array in ``arrs``.
+    arrs : list[np.ndarray] or np.ndarray
+        An array or a list of arrays to apply ``func`` to.
+    rowsize : list
+        List of integers specifying the number of data points in each row.
+    *args : tuple
+        Additional arguments to pass to ``func``.
+    max_workers : int, optional
+        Number of threads to use. If None, the number of threads will be equal
+        to the ``max_workers`` default value of ``concurrent.futures.ThreadPoolExecutor``.
+    **kwargs : dict
+        Additional keyword arguments to pass to ``func``.
+
+    Returns
+    -------
+    out : tuple[np.ndarray] or np.ndarray
+        Output array(s) from ``func``.
+
+    Examples
+    --------
+    >>> def func(x, y):
+    ...     return x + y
+    >>> x = np.arange(10)
+    >>> y = np.arange(10, 20)
+    >>> apply_ragged(func, [x, y], [5, 5])
+    array([10, 12, 14, 16, 18, 20, 22, 24, 26, 28])
+
+    Raises
+    ------
+    ValueError
+        If the sum of ``rowsize`` does not equal the length of ``arrs``.
+    """
+    # make sure the arrs is iterable
+    if not isinstance(arrs[0], Iterable):
```

Comment:
This fails if a DataArray is passed as a scalar because an element of a DataArray is also a DataArray (annoyingly so).